indexer/src/main.ts

```diff
@@ -0,0 +1,600 @@
+  if (!currentPosition) {
+    throw new Error('Position not found');
+  }
+  currentPosition.latest = false;
+  await ctx.store.upsert(currentPosition);
+
+  let collateral = currentPosition.collateralAmount;
+  const size = currentPosition.size - sizeDelta;
+  const realizedFundingRate = currentPosition.realizedFundingRate + fundingRate;
+  const realizedPnl = currentPosition.realizedPnl + pnlDelta;
+
+  collateral = collateral + fundingRate + pnlDelta - positionFee;
+  let collateralTransferred;
+  if (size === BigInt(0)) {
+    // close the position
+    collateralTransferred = collateral;
+    collateral = BigInt(0);
+  } else {
+    const collateralTarget = collateral - collateralDelta;
+    if (collateral > collateralTarget) {
+      collateralTransferred = collateral - collateralTarget;
+      collateral = collateralTarget;
+    } else {
+      collateralTransferred = BigInt(0);
+    }
+  }
+
+  const position: Position = new Position({
+    id: receipt.id,
+    positionKey,
+    collateralAmount: collateral,
+    size,
+    timestamp: getUTCBlockTime(block),
+    latest: true,
+    change: PositionChange.DECREASE,
+    collateralTransferred,
+    positionFee,
+    fundingRate,
+    pnlDelta,
+    realizedFundingRate,
+    realizedPnl,
+  });
+  await ctx.store.insert(position);
+
+  const totalPosition: TotalPosition | null = await ctx.store.findOne(TotalPosition, {
+    where: { indexAssetId: log.index_asset, isLong: log.is_long },
+  });
+  if (!totalPosition) {
+    throw new Error('Total position not found');
+  }
+  const collateralDiff = BigInt(currentPosition.collateralAmount) - collateral;
+  totalPosition.collateralAmount -= collateralDiff;
+  totalPosition.size -= sizeDelta;
+  totalPosition.lastTimestamp = getUTCBlockTime(block);
+  await ctx.store.upsert(totalPosition);
+}
+
+// The ClosePosition event goes right after the DecreasePosition event
+// so just update the status
+async function handleClosePosition(
+  receipt: Receipt<{ receipt: { rb: true; data: true } }>,
+  _block: Block,
+  ctx: any
+) {
+  const logs = vaultInterface.decodeLog(receipt.data!, receipt.rb!.toString());
+  const log = logs[0];
+  const positionKey = log.key;
+  const positionKeyRecord: PositionKey | null = await ctx.store.findOne(PositionKey, {
+    where: { id: positionKey },
+  });
+  if (!positionKeyRecord) {
+    throw new Error('Position key not found');
+  }
+  const currentPosition: Position | null = await ctx.store.findOne(Position, {
+    where: { positionKey: positionKeyRecord, latest: true },
+  });
+  if (!currentPosition) {
+    throw new Error('Position not found');
+  }
+  currentPosition.change = PositionChange.CLOSE;
+  await ctx.store.upsert(currentPosition);
+  // verify empty position
+}
+
+async function handleLiquidatePosition(
+  receipt: Receipt<{ receipt: { rb: true; data: true } }>,
+  block: Block,
+  ctx: any
+) {
+  const logs = vaultInterface.decodeLog(receipt.data!, receipt.rb!.toString());
+  const log = logs[0];
```

Comment:
We can't assume that's going to be the first log. What if it's a transaction that calls multiple contracts and each one emits an event?